keps/sig-instrumentation/0034-distributed-tracing-kep.md

```diff
@@ -81,55 +88,53 @@ Distributed tracing provides a single window into latency information from acros
+In the traditional tracing model, a client sends a request to a server and receives a response back.  The kube-apierver and backing storage (e.g. etcd3) follow this model, so we can easily add tracing for API Server requests.  To enable traces to be collected for API requests, the following must be true:
 
 1. The apiserver must propagate the http context of incoming requests through its function stack to the backing storage
 1. Kubernetes client libraries must allow passing a context with API requests
 
-To actually add traces to API requests, owners of the kube-apiserver and backing storage may add Spans to incoming requests, and configure sampling as they see fit.
+To add traces to API requests, owners of the kube-apiserver and backing storage may add Spans to incoming requests, and configure sampling as they see fit.
 
 #### Propagating Context Through Objects
 
-While API requests follow the traditional RPC client-server tracing model, kubernetes controllers don't.  Instead of controller actions being driven by incoming RPCs, their actions are driven by observations of desired and actual state.  This is the primary reason why the kubernetes community hasn't agreed on how to integrate tracing into kubernetes thus far.
+Unlike the traditional RPC model, controllers in kubernetes do not communicate directly with each other directly over HTTP.  Thus, we cannot use http headers to propagate the context between controllers.  Instead, controllers write to and read from objects.  To propagate the trace context from object writer to object reader, we will store the context in a field in ObjectMeta:
 
-In the traditional RPC client-server tracing model, a trace context is attached to a single incoming request, and is propagated with all requests the server makes to other servers required to fulfill the initial single request.  Conceptually, this proposal suggests treating a kubernetes cluster as a single RPC server.  The difference is that we attach context to objects, and propagate this context to objects modified as a result of the initial object modification.  For example, if a user creates a ReplicaSet, the kube-controller-manager will create many Pod objects as a result, and will propagate the context used to create the ReplicaSet to Pod objects as well.  This ensures that all actions taken by kubernetes controllers as a result of the initial user action are linked by the same context.
+```golang
+type ObjectMeta struct {
+  ...
+  // TraceContext is the portion of a Span in distributed tracing that is propagated between actors.
+  // Values written to this field must comply with the W3C standard for tracecontext: 
+  // https://www.w3.org/TR/trace-context/#tracestate-field
+  //
+  // +optional
+  TraceContext string `json:"traceContext,omitempty" protobuf:"bytes,18,opt,name=traceContext"``
```

Comment:
@liggitt 
Here is the proposed change to ObjectMeta.

Reply:
I'm not very excited about having this field here. Putting this in metadata blesses the concept in a way that having it in an annotation does not.

Here's a list of concerns:
* Unlimited lifespan. When does this stop being valid data?
* Linearization of an undirected, cyclic graph. Controller actions, even the built in ones, are not linearizable in the way implied by having a single trace context. Controllers fighting over an object's trace span will obscure more than illuminate.
* Half-mandatory. It doesn't seem possible for apiserver to force controllers to do the right thing with this field, so it seems highly likely that e.g. some controllers (probably the ones you're most interested in debugging) will not use it.
* Not obvious whether this is for reading or writing (likely could be fixed with better documentation). I am understanding this field as documenting, for the next controller, a span to propagate. Having it here also makes it appear like apiserver is responsible for acting on this value, when it's actually up to the controllers to report their actions (?).

I also don't understand: why do you need to set the span when writing status?

I'll leave another comment with a different design approach.

Reply:
Even though the KEP no longer proposes a field, i'll still try and address the comments above.

> Unlimited lifespan

It continues to be valid data forever.  It would be helpful to have a mechanism to limit this, which I added as an item for beta.

> Half-mandatory

The new proposal enforces behavior with a mutating admission controller, and simplifies the controller implementation to ensure doing the right thing is easy.

> Reading or writing?

I've separated the proposal into a Reading (i.e. choosing a parent) and Writing sections to make this clearer.

> Linearization of an undirected, cyclic graph

Controller actions do not form a cyclic graph.  Time is linear, and each action must specify a parent action (i.e. span) that was created before it.

> why do you need to set the span when writing status?

Removed in the latest iteration of the proposal.  It was to address the "unlimited lifespan" problem, but I think there are better ways to solve it that don't involve writing the span with status.